### PR TITLE
Initial Core Implementation of Lingo.dev Go SDK

### LINKED ISSUE #1073: Create Official Lingo.dev GO SDK
## Overview
Create an official GO SDK for Lingo.dev API to complement existing SDKs (JavaScript, PHP, Python).

## Requirements
- Reference existing SDKs for consistency:
  - JavaScript SDK (exclude HTML processing - keep that JS-only)
  - PHP SDK 
  - Python SDK
- Support exponential backoff for retries
- Support chunking for large requests (same as JS SDK)
- Comprehensive unit tests
- Integration tests requiring LINGODOTDEV_API_KEY environment variable
- Follow Go best practices and conventions
- Proper error handling and logging
- Documentation and examples

## Deliverables
- [ ] Core SDK implementation
- [ ] Unit test suite
- [ ] Integration test suite
- [ ] Documentation
- [ ] Usage examples
- [ ] CI/CD pipeline integration

## Testing Strategy
- Unit tests: Mock API responses, test core functionality
- Integration tests: Real API calls with valid API key, test end-to-end workflows

## Technical Notes
- Must maintain feature parity with existing SDKs (minus HTML processing)
- Follow semantic versioning
- Include proper Go module structure

### PR DESCRIPTION
**Pull request by iYashMaurya**
## Overview
This PR delivers the complete, production-ready **Lingo.dev Go SDK**. It achieves full feature parity with the existing Python, JavaScript, and PHP SDKs (excluding HTML processing, which remains JS-only). 

The SDK has been built from the ground up to follow Go best practices, featuring robust concurrency, context propagation, comprehensive error handling, and fully automated CI/CD pipelines.

Closes https://github.com/lingodotdev/lingo.dev/issues/1073

### Core SDK Features
- **Full API Surface:** Implemented `LocalizeText`, `LocalizeObject`, `LocalizeChat`, `BatchLocalizeText`, `BatchLocalizeObjects`, `RecognizeLocale`, and `WhoAmI`.
- **First-Class Context Support:** Every method accepts `context.Context` as the first parameter, propagating all the way down to `http.NewRequestWithContext` for reliable request cancellation.
- **Resilience:** Automatic exponential backoff (1s, 2s, 4s) on 5xx server errors and network failures (4xx errors fail fast).
- **Concurrent Chunk Processing:** Large payloads are automatically split into chunks and processed in parallel using `golang.org/x/sync/errgroup`.
- **Typed Error Handling:** Introduced `ValueError` for validation/400 errors and `RuntimeError` (with a `StatusCode` field) for server/network errors.
- **Flexible Configuration:** Utilizes functional options (`SetURL`, `SetBatchSize`, `SetIdealBatchItemSize`) for clean client initialization.

### 🧪 Testing (82.6% Coverage)
- **Unit Tests (30 tests):** Comprehensive testing using `httptest.NewServer` covering config validation, chunking, truncation, error types, and core endpoints.
- **Integration Tests (9 tests):** Real API verification against production endpoints. (Automatically skipped when `LINGODOTDEV_API_KEY` is not set).
- **Mocked Integration Tests (7 tests):** Validates complex internal logic including retry behavior, concurrent parallelism, context cancellation, and workflow ID consistency.
- **Performance:** Unit + Mocked tests complete in ~13s; Full integration suite completes in ~60s.

### ⚙️ CI/CD & Automation
- **PR Workflow:** Matrix testing across Go 1.21 and 1.22. Includes build, `vet`, `gofmt` checks, conditional integration tests, and coverage uploads to Codecov.
- **Release Workflow:** Automated semantic versioning driven by conventional commits, automated changelog generation via `git-chglog`, and GitHub release creation.
- **Standards:** Added a PR template enforcing conventional commit guidelines.

### 📚 Documentation
- **Comprehensive README:** Includes CI badges, runnable examples for every method, a configuration table, full API references, and sections on error handling and Go-specific advantages.
- **Developer Experience:** Added a fully runnable `examples/main.go` and a `Makefile` with targets for `test`, `test-integration`, `test-coverage`, `build`, `vet`, `fmt`, `lint`, and `clean`.

## 🏗️ Architecture & Request Flow

```text
Client.LocalizeText(ctx, ...)
  → localizeRaw(ctx, ...)
    → ExtractChunks(payload)
    → localizeChunk(ctx, ...) [sequential or concurrent via errgroup]
      → do(ctx, ...) [retries, JSON parsing, returns full response]
        → http.NewRequestWithContext(ctx, ...)
 ```
 ## 📝 Technical Notes for Reviewers
- **Context Flow:** Canceling the provided context will successfully abort in-flight HTTP requests natively. *(Note: Retry logic sleeps between attempts but does not currently check context during the sleep timer).*
- **Unified HTTP Layer:** The internal `do()` method returns a `map[string]any` representing the full JSON response. This allows divergent response structures (like `/recognize` vs `/i18n`) to pass through the same robust HTTP and retry layer. Callers extract exactly what they need.
- **Concurrency:** Batch methods rely on `errgroup.WithContext` to manage parallel execution, ensuring automatic cancellation of sibling routines if one fails.

## 📦 Dependencies
- `golang.org/x/sync` — Utilized `errgroup` for safe concurrent chunk and batch processing.
- `github.com/matoous/go-nanoid/v2` — Used for robust workflow ID generation.

## Summary by CodeRabbit

* **New Features**
  * Official Go SDK: client constructor, localization APIs (text, objects, chat), batch multi‑locale operations, locale recognition, identity lookup, typed errors, retries, response truncation, and configurable batching/concurrency options.

* **Documentation**
  * Expanded README, changelog and changelog template/config, examples, and a PR template.

* **Tests**
  * Comprehensive unit and integration suites (mocked and real‑API), covering chunking, retries, concurrency, and edge cases.

* **Chores**
  * CI/workflows, release automation, Makefile, license, gitignore, and module initialization.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	total := len(payload)
+	processed := 0
+	var result []map[string]any
+	currentChunk := make(map[string]any)
+	var currentItemCount int
+
+	for key, value := range payload {
+		currentChunk[key] = value
+		currentItemCount++
+		currentChunkSize := CountWords(currentChunk)
```
Comment (cherkanovart) [inline client.go]: This recalculates CountWords across the whole currentChunk on every iteration. For a chunk of 250 items with nested values we end up walking the same tree 250 times, which makes chunking effectively O(n^2) on the total payload.

Quick fix is to track the running count incrementally:

```go
currentChunk[key] = value
currentWordCount += CountWords(value)
currentItemCount++
```

and reset both counters when we flush a chunk.

Small bonus thing: the `processed == total` check mixes "we hit a limit" with "we ran out of items". If you go incremental you can drop that condition and just flush whatever is left after the loop:

```go
if len(currentChunk) > 0 {
    result = append(result, currentChunk)
}
```

File: engine.go [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+	fast := false
+	if params.Fast != nil {
+		fast = *params.Fast
+	}
+
+	merged := make(map[string]any)
+
+	if concurrent {
+		var mu sync.Mutex
+		g, gCtx := errgroup.WithContext(ctx)
```
Comment (cherkanovart) [inline engine.go:37]: Heads up, this fires a goroutine per chunk with no upper bound. If someone calls LocalizeObject with a few MB of strings we can end up with hundreds of in-flight requests against engine.lingo.dev at once, and once the API starts returning 429s the user just sees the SDK hang.

Easiest fix is `g.SetLimit(...)` right after `errgroup.WithContext`. Probably worth adding a `MaxConcurrency int` to Config with a default like 10, plus a `SetMaxConcurrency` option so users can tune it.

Same pattern needs to land in BatchLocalizeText and BatchLocalizeObjects below.

File: engine.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	chatPayload := make([]any, len(chat))
+	for i, msg := range chat {
+		chatPayload[i] = map[string]any{
+			"name": msg["name"],
+			"text": msg["text"],
+		}
+	}
+	payload := map[string]any{"chat": chatPayload}
```
Comment (cherkanovart) [inline engine.go]: This wraps the whole chat array as a single top-level key, then sends it through localizeRaw which chunks by iterating top-level map keys. Since there is only one key (`chat`) here, the entire conversation always goes out as a single HTTP request no matter how long it is.

For a 500 message chat that becomes one giant payload that will very likely hit the 60s http client timeout, and the chunking the SDK promises never actually kicks in for chat.

Suggestion: chunk the chat slice in LocalizeChat itself before sending, something like:

```go
var subSlices [][]any
for i := 0; i < len(chatPayload); i += c.config.BatchSize {
    end := i + c.config.BatchSize
    if end > len(chatPayload) {
        end = len(chatPayload)
    }
    subSlices = append(subSlices, chatPayload[i:end])
}
```

then send each subSlice through localizeChunk concurrently (same workflowID, errgroup with SetLimit), and concat the results in order.

Would be great to also add a test where BatchSize=2 and the chat has 5 messages, asserting the mock server actually saw multiple requests.

File: engine.go [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+func (c *Client) BatchLocalizeText(ctx context.Context, text string, sourceLocale *string, fast *bool, targetLocales []string) ([]string, error) {
+	if text == "" {
+		return nil, &ValueError{Message: "lingo: text must not be empty"}
+	}
+	if len(targetLocales) == 0 {
+		return []string{}, nil
+	}
+
+	results := make([]string, len(targetLocales))
+	g, gCtx := errgroup.WithContext(ctx)
```
Comment (cherkanovart) [inline engine.go:341]: Same unbounded fan-out as localizeRaw. If a user passes 50 target locales that is 50 simultaneous requests. Needs `g.SetLimit(c.config.MaxConcurrency)` too.

File: engine.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+// BatchLocalizeObjects translates multiple objects concurrently using the same localization params.
+func (c *Client) BatchLocalizeObjects(ctx context.Context, objects []map[string]any, params LocalizationParams) ([]map[string]any, error) {
+	if len(objects) == 0 {
+		return []map[string]any{}, nil
+	}
+
+	results := make([]map[string]any, len(objects))
+	g, gCtx := errgroup.WithContext(ctx)
```
Comment (cherkanovart) [inline engine.go:375]: And same here, `g.SetLimit` needed.

**Review by cherkanovart — COMMENTED** [review]
Hey, really nice work on this PR overall, the structure and the test coverage are solid. Did a closer read and found a few things worth fixing before merging, mostly around how the SDK behaves on larger payloads. Comments inline.

**Comment by iYashMaurya** [comment]
Hey @cherkanovart ,
I will be addressing all this issue and pushing the new version by end of week if it's ok?
